### PR TITLE
Feature/reporters object model fix

### PR DESCRIPTION
**Pull request by pesse**
This will break cli and java-api
Don't merge before https://github.com/utPLSQL/utPLSQL-java-api/pull/45 is finished (I'm actually working on it)

`ut_output_buffer` is now fully abstracted from caller and caller can…

… use reporter to get the output data.

Introduced new abstract type: `ut_output_reporter_base`.
Every reporter based on that type is supposed to use output buffer.
So for every reporter of type `ut_output_reporter_base`, client can call functions:
- get_lines
- lines_to_dbms_output
- get_lines_cursor

Also, the reporters don't need to call parent reporter's method to finalize the reporting.
New finalize event is added.
There is no direct dependency between `ut_runner.run` and `ut_output_buffer` so we may now pass a reporter that doesn't report to output but saves data to table for example.

Removed unused code. Added unit tests for better coverage.

Refactored `ut_output_buffer` package into `ut_output_table_buffer` type.
Created base type for output_buffers.
Created header table for table output data.
Updated `ut_output_reporter_base` to contain by default `ut_output_table_buffer`
Updated tests for output buffers.
Refreshed list of sources for RunAll script.

Addressed review comments and fixed tests stability.

Removing initialization from output constructor

Fixed buffer initialization.

Fixed tests location after rebase.

### CODE REVIEW

**Comment by jgebal** [comment]
Let me know when you're good to go on java

**Comment by pesse** [comment]
Closes #592

**Comment by jgebal** [comment]
This PR will break java-api, so we cannot merge it until changes are implemented on java side.

File: source/reporters/ut_coverage_sonar_reporter.tpb [hunk 1]
```diff
@@ -86,8 +86,12 @@ create or replace type body ut_coverage_sonar_reporter is
     l_coverage_data := ut_coverage.get_coverage_data(a_run.coverage_options);
 
     self.print_clob( get_coverage_xml( l_coverage_data ) );
+  end;
 
-    (self as ut_reporter_base).after_calling_run(a_run);
+  overriding member function get_description return varchar2 as
+  begin
+    return 'Generates a JSON coverage report providing information on code coverage with line numbers.' || chr(10) ||
+           'Designed for [SonarQube](https://about.sonarqube.com/) to report coverage.';
```
Comment (jgebal) [inline source/reporters/ut_coverage_sonar_reporter.tpb]: I would add line:
`JSON format returned conforms with the Sonar specification: https://docs.sonarqube.org/display/SONAR/Generic+Test+Data`

File: source/reporters/ut_coveralls_reporter.tpb [hunk 2]
```diff
@@ -97,8 +97,12 @@ create or replace type body ut_coveralls_reporter is
     l_coverage_data := ut_coverage.get_coverage_data(a_run.coverage_options);
 
     self.print_clob( get_coverage_json( l_coverage_data ) );
+  end;
 
-    (self as ut_reporter_base).after_calling_run(a_run);
+  overriding member function get_description return varchar2 as
+  begin
+    return 'Generates a JSON coverage report providing information on code coverage with line numbers.' || chr(10) ||
+           'Designed for [Coveralls](https://coveralls.io/).';
```
Comment (jgebal) [inline source/reporters/ut_coveralls_reporter.tpb]: I would add link:
`JSON format conforms with specification: https://docs.coveralls.io/api-introduction`

File: source/reporters/ut_sonar_test_reporter.tpb [hunk 3]
```diff
@@ -97,5 +97,11 @@ create or replace type body ut_sonar_test_reporter is
     self.print_text('</testExecutions>');
   end;
 
+  overriding member function get_description return varchar2 as
+  begin
+    return 'Generates a JSON report providing detailed information on test execution.' || chr(10) ||
```
Comment (jgebal) [inline source/reporters/ut_sonar_test_reporter.tpb]: I would add line:
`JSON format returned conforms with the Sonar specification: https://docs.sonarqube.org/display/SONAR/Generic+Test+Data`

File: source/reporters/ut_teamcity_reporter.tpb [hunk 4]
```diff
@@ -113,5 +113,10 @@ create or replace type body ut_teamcity_reporter is
 
   end;
 
+  overriding member function get_description return varchar2 as
+  begin
+    return 'For reporting live progress of test execution with Teamcity CI.';
```
Comment (jgebal) [inline source/reporters/ut_teamcity_reporter.tpb]: As described in documentation: `https://confluence.jetbrains.com/display/TCD9/Build+Script+Interaction+with+TeamCity`

File: source/reporters/ut_xunit_reporter.tpb [hunk 5]
```diff
@@ -130,5 +129,10 @@ create or replace type body ut_xunit_reporter is
            || '" time="' || ut_utils.to_xml_number_format(a_item.execution_time()) || '" ';
   end;
 
+  overriding member function get_description return varchar2 as
+  begin
+    return 'Used for reporting test results with CI servers like Jenkins/Hudson/Teamcity.';
```
Comment (jgebal) [inline source/reporters/ut_xunit_reporter.tpb]: `Provides outcomes in a format conforming with JUnit 4 and above as defined in:
 https://gist.github.com/kuzuha/232902acab1344d6b578`

File: tests/ut_expectations/ut.expect.not_to_equal.anydata.GivesFailureWhenComparingTheSameData.sql [hunk 6]
```diff
@@ -0,0 +1,17 @@
+--Arrange
```
Comment (jgebal) [inline tests/ut_expectations/ut.expect.not_to_equal.anydata.GivesFailureWhenComparingTheSameData.sql]: Those tests should be moved to utPLSQL v3 test pacakges.

File: old_tests/RunAll.sql [hunk 7]
```diff
@@ -72,7 +72,8 @@ exec ut_coverage.coverage_start_develop();
 @@lib/RunTest.sql ut_metadata/ut_metadata.form_name.TrimStandaloneProgramName.sql
 
 @@lib/RunTest.sql ut_output_buffer/get_lines.RecievesALineFromBufferTableAndDeletes.sql
-@@lib/RunTest.sql ut_output_buffer/send_line.DoesNotSendLineIfNullReporterIdGiven.sql
+@@lib/RunTest.sql ut_output_buffer/get_lines.WaitsForMoreDataToAppearForSpecifiedTime.sql
```
[the remark this answers is not in the document]
Comment (pesse) [reply old_tests/RunAll.sql]: I'll try to, but will work on java-api before. That's something we can change after 3.1.0 as part of refactoring, can't we?

**Pull request merged**